Fix KMS encryption script to run in GCP cloud shell

Edwin-Pau opened the pull request:
Applied several fixes that enable the KMS encryption script under tools to be run from GCP cloud shell directly. A use-case would be if a user wanted to use the script to decrypt the terraform.tfvars file after completing the GCP quickstart deployment.

Signed-off-by: Edwin-Pau epau@teradici.com

Commit:
GCP: Fix module import error for encryption script

Added a function import_or_install_module(pip_package, module_name)
which attempts to import a specified Python package, and if it fails
it will attempt to install it first before importing it again.

This function takes in two arguments: the first one is the Python
package name and the second optional parameter is for the import name
(if the import name is different than the package name, e.g.
"google-cloud-kms" for pip3, "google.cloud.kms_v1" for import.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Use absolute paths in quickstart

Modified the GCP quickstart script to pass in absolute paths instead
of relative paths for the following terraform.tfvars variables:
gcp_credentials_file, cac_admin_ssh_pub_key_file and
centos_admin_ssh_pub_key_file.

This change allows the KMS encryption script to work in both the local
and cloud environments. This is also in-line with the terraform.tfvars
.sample file.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
Update docs for absolute paths

Updated both the GCP and AWS docs under the "Customizing
terraform.tfvars" section to explicitly state that absolute paths
should be used instead of implicitly through the examples shown.

Fixed a typo under the GCP quickstart tutorial.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix encryption script shebang line

Fixed the shebang line for the KMS encryption script to use /usr/bin/env
which uses the user's path to run the script, making it more portable
to different environments, such as in the Google Cloud Shell.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Use f-strings instead of str.format()

Replaced all usage of str.format() with f-strings instead in the
KMS encryption script.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Move error string to SystemExit()

Moved error strings to SystemExit() calls instead of using print
statements.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix quickstart script

Fixed GCP quickstart shebang line to use /usr/bin/env
which uses the user's path to run the script, making it more portable
to different environments.

Grouped the current working directory variable (cwd) with the other
os function calls.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Remove static imports

Removed static importing of Google modules as there's a custom function
which handles the install and imports of these Google modules.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Edwin-Pau commented:
Please refer to the closed PR #112 for revision comments.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -7,13 +7,48 @@
+        print(f"Successfully imported {module_name}.")
+
+    except ImportError:
+        install_cmd = f"pip3 install {pip_package} --user"
```
Can we add a prompt for user confirmation? Something like: "this script requires <pip-package-name> but is not available.  Install package by running "pip3 install {pip-package-name} --user"? (Y/N)" and if user enters N then quit.

Commit:
GCP: Remove sudo in quickstart script

Removed the use of 'sudo' when installing Google PIP modules and
Terraform.

Changed the Terraform binary install directory to use '$HOME/bin'.

Added checks and prompts user before installing Terraform and Google
PIP modules.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Add cleanup instructions for quickstart

If Terraform was installed using the quickstart script, a different
command must be run to tear down the deployment.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Update quickstart clean up message

Updated the GCP quickstart clean up message with the correct
directory and a different command for cleaning up using Terraform.

Signed-off-by: Edwin-Pau <epau@teradici.com>

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -101,6 +104,25 @@ def check_requirements():
         print('')
         sys.exit(1)
 
+    path = shutil.which('terraform')
```
where you check for terraform and the required python packages are in different functions, which seems a bit confusing.  I'd address the requirements in this function.  maybe rename the function to "ensure_requirements" instead of "check_requirements".  so 1) check for different requirements, if work is needed then ask for permission and set a flag to true.  we want to ask all the questions at the beginning of the script.  2) at the end of this function, go thru the flags and install the missing stuff.

shermanyin replied:
Also need to check terraform version > 0.12.  If not then install a compatible version.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -101,6 +104,25 @@ def check_requirements():
+        from google.api_core import exceptions as google_exc
+
+    except ImportError as err:
+        install_permission = input("Google Cloud module is not installed, proceed with installation (y/N)? ").strip().lower()
```
"Google Cloud module" sounds a bit weird.  I think those should be called "Python Google Client Libraries".  Same for other references to "modules"

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -274,16 +296,29 @@ def tf_vars_create(ref_file_path, tfvar_file_path, settings):
-        install_cmd = 'sudo python3 install-terraform.py'
-        subprocess.run(install_cmd.split(' '), check=True)
 
+    path = shutil.which('terraform')
```
no need for this check again if we move this code to under check_requirements() after the check for terraform

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -101,6 +104,25 @@ def check_requirements():
+        if install_permission not in ('y', 'yes'):
+            print('Terraform is required for deployment. Exiting...')
+            sys.exit(1)
+
```
add a variable here like "install_terraform = true" and use that to determine if we need to perform installation at end of this function.

Edwin-Pau replied:
I've made this flag a global one, since we want to reference it as part of the next_steps string.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -101,6 +104,25 @@ def check_requirements():
+        if install_permission not in ('y', 'yes'):
+            print('Google Cloud modules are required for deployment. Exiting...')
+            sys.exit(1)
+
```
here add a flag that like install_google_libraries = true

Then go thru the flags and install missing stuff

Edwin-Pau replied:
For consistency sake, I've made this flag global as well.

Edwin-Pau replied:
Flags have been removed since the logic to handle installing Terraform or Google libraries are now in their own functions.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -90,8 +90,10 @@
+  2. In GCP cloudshell, change directory using the command "cd ~/cloudshell_open/cloud_deployment_scripts/{deployment_path}"
   3. Remove resources deployed by Terraform using the command "terraform destroy". Enter "yes" when prompted.
+     Note: If Terraform was installed using this quickstart script, execute the following command instead.
+     "~/bin/terraform destroy"
```
not need for "if" here right? we know if we installed the script or not. Just display the proper command based on whether or not we installed terraform?

Commit:
GCP: Ensure requirements met for quickstart

Added two functions which ensure the requirements are met before
proceeding with the quickstart deployment.

Moved all the logic to check and install Terraform into the new
function renamed ensure_terraform(). Moved all the logic to check
and install Google libraries into the new function renamed
ensure_google_libraries().

If they are not installed, the script prompts the user
for permission to install the requirements for both Terraform and
Google libraries.

Added Terraform version checking.

Fixed the correct Python packages to check for and install.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Edwin-Pau commented:
- Rebased on staging due to merge conflicts.
- Tested on Google Cloud Shell without Terraform installed.
- Tested on local machine without Google libraries installed.

Edwin-Pau commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -298,40 +359,20 @@ def tf_vars_create(ref_file_path, tfvar_file_path, settings):
-    check_requirements()
+    ensure_requirements()
+
+    import googleapiclient.discovery
+    from google.cloud import kms_v1
+    from google.cloud.kms_v1 import enums
+    from google.api_core import exceptions as google_exc
+    print('Successfully imported Google libraries.\n')
```
We need to import in main(), otherwise it can't find use the import references, unless we import and return modules in main, like in the encryption script. But that would be part of a future task which is refactoring this script to use the encryption script.
`Encrypting secrets...
Traceback (most recent call last):
  File "./gcp-cloudshell-quickstart.py", line 418, in <module>
    kms_client = kms_v1.KeyManagementServiceClient()
NameError: name 'kms_v1' is not defined`

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -7,13 +7,61 @@
+    If the required package is not installed, it will install the package before importing it again.
+
+    Args:
+        pip_package_name (str): the name of the pip package to be installed
```
should be called pypi package.

shermanyin commented on quickstart/install-terraform.py:
```diff
@@ -3,18 +3,24 @@
 
+
 def terraform_install(version=TERRAFORM_VERSION):
+
```
empty line not needed.

shermanyin commented on quickstart/install-terraform.py:
```diff
@@ -35,14 +42,19 @@ def terraform_install(version=TERRAFORM_VERSION):
+    parser.add_argument('dir', help='specify the directory to install Terraform')
+    args = parser.parse_args()
+
+    TERRAFORM_BIN_DIR = args.dir
```
so this overrides L18?

Edwin-Pau replied:
Yes, the command line argument would override the global bin dir. The script's default is the same as what we pass in at the moment, is this ok?

shermanyin replied:
Is there a scenario when the value assigned in L18 is used then?

Edwin-Pau replied:
Currently L18 would not be used - should it be initialized to None instead?

Or should I make the argparse directory argument an optional one? That way if no argument is specified to insta--terraform.py, then it would use the default directory in the script. Which way is better?

shermanyin replied:
How about remove the global, make it a required parameter for this file, and pass it into the terraform_install() function as a required parameter.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -40,11 +40,19 @@
 CONNECTOR_NAME  = 'quickstart_cac_' + iso_time
 
+# Install flags
+INSTALL_TERRAFORM        = False
```
why do these have to be global if they are only used within one function?

Edwin-Pau replied:
`INSTALL_TERRAFORM` is used towards the end to print the command string back to the user. I guess leaving the fully qualified path regardless of Terraform is installed or not would work too right? But I just thought it may be cleaner for the user if they can just use `terraform destroy` vs `/home/bin/terraform destroy`

Edwin-Pau replied:
Global flags have been removed.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -119,18 +217,18 @@ def quickstart_config_read(cfg_file):
-       - unicode characters
-
-    See: https://docs.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/password-must-meet-complexity-requirements
+Please enter a password for the Active Directory Administrator.
```
use this instead? https://docs.python.org/3/library/textwrap.html#textwrap.dedent

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -88,20 +96,110 @@
+
+    global INSTALL_GOOGLE_LIBRARIES
+
+    google_libraries_import = """
```
https://docs.python.org/3/library/textwrap.html#textwrap.dedent

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -88,20 +96,110 @@
+        
+        print(f'Found {terraform_version} in {path}.\n')
+
+        if 'v0.12.' not in terraform_version:
```
You can check if v0.12. is in terraform_version, if so then just return the path.

Then you don't need the else and just proceed to install terrform without an extra if on L191.

Also returning the path will mean you avoid referencing the global variable.

Commit:
GCP: Fix quickstart script

-Removed global install flags in quickstart
-Added textwrap.dedent to improve multi-line strings
-Fixed the function that checks for v.12. before installing Terraform
-Fixed the next_steps string to display the correct Terraform path

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix encryption script

-Changed pip_package_name to pypi_package_name
-Used sys.executable instead of python3

Signed-off-by: Edwin-Pau <epau@teradici.com>

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -7,13 +7,61 @@
+
+    Args:
+        pypi_package_name (str): the name of the PyPI package to be installed
+        module_name (str): the import name of the module if it is different than the pip package name
```
not "pip package"

Commit:
GCP: Fix Terraform install script

The install script for Terraform has been optimized by removing
unnecessary global variables and it now simply installs Terraform
into the user specified directory from the command line argument.

The logic to build the path to install Terraform in was redundant
since the GCP quickstart script already handles this.

Signed-off-by: Edwin-Pau <epau@teradici.com>

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -124,15 +212,15 @@ def ad_password_get():
-       - 0-9
-       - special characters: (~!@#$%^&*_-+=`|\(){}[]:;"'<>,.?/)
-       - unicode characters
+    - A-Z
+    - a-z
+    - 0-9
+    - special characters: (~!@#$%^&*_-+=`|\(){}[]:;"'<>,.?/)
+    - unicode characters
```
don't need to get rid of these spaces

shermanyin commented on quickstart/install-terraform.py:
```diff
@@ -3,30 +3,33 @@
 
-def terraform_install(version=TERRAFORM_VERSION):
+
+def terraform_install(terraform_bin_dir, version=TERRAFORM_VERSION):
```
the function would make more sense if "version" is optional.  If it's not specified, install latest.  otherwise, install the specific version.  but let's leave this for a separate PR.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -88,20 +93,103 @@
+        
+        print(f'Found {terraform_version} in {path}.')
+
+        if 'v0.12.' in terraform_version:
```
this doesn't actually check for >=0.12.  For example, 0.13 would fail this check.

Edwin-Pau commented:
These changes have been pushed and merged from BitBucket, closing this PR.
